loadbalancer/algorithm.go

```diff
@@ -190,31 +191,51 @@ func (r *random) Apply(ctx *routing.LBContext) routing.LBEndpoint {
 	return withFadeIn(r.rand, ctx, r.notFadingIndexes, r.fadingWeights, i)
 }
 
-type consistentHash struct{}
+type (
+	endpointHash struct {
+		index int    // index of endpoint in endpoint list
+		hash  uint32 // hash of endpoint
+	}
+	consistentHash []endpointHash // list of endpoints sorted by hash value
+)
+
+func (ch consistentHash) Len() int           { return len(ch) }
+func (ch consistentHash) Less(i, j int) bool { return ch[i].hash < ch[j].hash }
+func (ch consistentHash) Swap(i, j int)      { ch[i], ch[j] = ch[j], ch[i] }
 
 func newConsistentHash(endpoints []string) routing.LBAlgorithm {
-	return &consistentHash{}
+	ch := consistentHash(make([]endpointHash, len(endpoints)))
+	for i, ep := range endpoints {
+		ch[i] = endpointHash{i, hash(ep)}
+	}
+	sort.Sort(ch)
+	return ch
+}
+
+func hash(s string) uint32 {
+	h := fnv.New32a()
+	h.Write([]byte(s))
+	return h.Sum32()
+}
+
+// Returns index of endpoint with closest hash to key's hash
+func (ch consistentHash) search(key string) int {
+	h := hash(key)
+	i := sort.Search(ch.Len(), func(i int) bool { return ch[i].hash >= h })
+	if i == ch.Len() { // rollover
```

Comment:
could be simplified to:
`i := sort.Search(..) % ch.Len()`

Reply:
Yes, this will also work, here I explicitly check for the special case in line with binary search [documentation](https://golang.org/pkg/sort/#Search):
> ... Search returns the first true index. If there is no such index, Search returns n.

> A common use of Search is to find the index i for a value x in a sorted, indexable data structure ...
> ... i is the index where it would be inserted.

Reply:
As far as I understand `i` can't be `>= ch.Len()`, so the "rollover" can be done without comparison using modulo operation.
I am fine with the +3 lines, but for me it seems more natural to use modulo.